File: day04/day04.cs
using System;
using System.Linq;
using Serilog;

namespace Shunty.AdventOfCode2019
{
    public class Day04 : IAoCRunner
    {
        public void Run(ILogger log)
        {
            // Puzzle input
            const int Lower = 372304;
            const int Upper = 847060;

            int part1 = 0, part2 = 0;
            var range = Enumerable.Range(Lower, Upper - Lower);
            foreach (var num in range)
            {
                var pwd = num.ToString();

                bool incrementing = true, haspair = false, hasdistinctpair = false;
                for (var index = 1; index < 6; index++)
                {
                    var left = pwd[index - 1];
                    var right = pwd[index];

                    // Check incrementing digits
                    if (right < left)
                    {
                        incrementing = false;
                        break;
                    }

                    // Check that L/R pair do not have neighbours that are the same
                    if (left == right && !hasdistinctpair)
                    {
                        haspair = true;
                        if (index > 1)
                        {
                            var leftneighbour = pwd[index - 2];
                            if (leftneighbour == left)
                                continue;
                        }
                        if (index < 5)
                        {
                            var rightneighbour = pwd[index + 1];
                            if (rightneighbour == right)
                                continue;
                        }
                        hasdistinctpair = true;
                    }
                }

                part1 += incrementing && haspair ? 1 : 0;
                part2 += incrementing && hasdistinctpair ? 1 : 0;
            }

            Console.WriteLine($"Part 1: {part1}");
            Console.WriteLine($"Part 2: {part2}");
        }
    }
}

File: day04/day04.go
package main

import "fmt"

func main() {
	// Puzzle input
	const Lower, Upper = 372304, 847060

	var part1, part2, plen = 0, 0, 6
	for i := Lower; i <= Upper; i++ {
		incrementing, haspair, hasdistinctpair := true, false, false
		pwd := fmt.Sprintf("%d", i)

		for i := 1; i < plen; i++ { // We know all the characters are 0..9 so we can ignore UTF-8/Unicode issues
			lch := pwd[i-1]
			rch := pwd[i]

			if rch < lch {
				// Fail. Not incrementing
				incrementing = false
				break
			}

			if lch == rch {
				haspair = true

				// Check if the left or right neighbours (if any) are the
				// same character. Skip if so.
				if i > 1 {
					lneigh := pwd[i-2]
					if lneigh == lch {
						continue
					}
				}
				if i < plen-1 {
					rneigh := pwd[i+1]
					if rneigh == rch {
						continue
					}
				}
				hasdistinctpair = true
			}
		}
		if incrementing && haspair {
			part1++
		}
		if incrementing && hasdistinctpair {
			part2++
		}
	}

	fmt.Println("Part 1: ", part1)
	fmt.Println("Part 2: ", part2)
}
